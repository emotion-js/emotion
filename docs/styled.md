---
title: "Styled Components"
---

`styled` is a way to create React or Preact components that have styles attached to them. It's available from [react-emotion](/packages/react-emotion) and [preact-emotion](/packages/preact-emotion). `styled` was heavily inspired by [styled-components](https://www.styled-components.com/) and [glamorous](https://glamorous.rocks/)

### Styling elements and components

`styled` is very similar to `css` except you call it with an html tag or React/Preact component and then call that with a template literal for string styles or a regular function call for object styles.

```jsx live
import styled from 'react-emotion'

const Button = styled('button')`
  color: turquoise;
`

render(<Button>This my button component.</Button>)
```

### Changing based on props

Any interpolations or arguments that are functions in `styled` are called with `props`, this allows you to change the styles of a component based on the props.

```jsx live
import styled from 'react-emotion'

const Button = styled('button')`
  color: ${props =>
    props.primary ? 'hotpink' : 'turquoise'};
`

const Container = styled('div')(props => ({
  display: 'flex',
  flexDirection: props.column && 'column'
}))

render(
  <Container column>
    <Button>
      This is a regular button.
    </Button>
    <Button primary>
      This is a primary button.
    </Button>
  </Container>
)
```

### Styling any component

`styled` can style any component as long as it accepts a `className` prop.

```jsx live
import styled from 'react-emotion'
const Basic = ({ className }) => (
  <div className={className}>Some text</div>
)

const Fancy = styled(Basic)`
  color: hotpink;
`

render(<Fancy />)
```


### Change the rendered tag using `withComponent`

Sometimes you want to create some styles with one component but then use those styles again with another component, the `withComponent` method can be used for this. This was inspired by [styled-components' `withComponent`](https://www.styled-components.com/docs/api#withcomponent).

```jsx live
// Create a section element
const Section = styled('section')`
  background: #333;
`
// Create an aside element with the same styles as Section
const Aside = Section.withComponent('aside')
render(
  <div>
    <Section>This is a section</Section>
    <Aside>This is an an aside</Aside>
  </div>
)
```

### Targeting another emotion component

Similar to [styled-components](https://www.styled-components.com/docs/faqs#can-i-refer-to-other-components), emotion allows for emotion components to be targeted like regular CSS selectors when using [babel-plugin-emotion](/packages/babel-plugin-emotion.md).

```jsx live
const Child = styled('div')`
  color: red;
`

const Parent = styled('div')`
  ${Child} {
    color: green;
  }
`
render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

Component selectors can also be used with object styles.

```jsx live
const Child = styled('div')({
  color: 'red'
})

const Parent = styled('div')({
  [Child]: {
    color: 'green'
  }
})

render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

### Pass refs down using `innerRef`

Sometimes you need to get a [ref](https://reactjs.org/docs/refs-and-the-dom.html) but passing `ref` to a styled component will return a ref to the styled component, not the component that it renders which is generally the one you want. You can pass `innerRef` instead of `ref` to get the ref of the component that styled renders.

```jsx live
const Input = styled('input')`
  color: hotpink;
`

const Button = styled('button')`
  color: green;
`

function TextInput(props) {
  let textInput = null
  function handleClick() {
    textInput.focus()
  }
  return (
    <div>
      <Input
        innerRef={input => {
          textInput = input
        }}
      />
      <Button onClick={handleClick}>
        Focus the text input
      </Button>
    </div>
  )
}
render(<TextInput />)
```

### Element Shorthand

> Note:

> `babel-plugin-emotion` is required for the element shorthand

Instead of using the function call syntax(`styled('div')`), you can use create components by using a property, where the property refers to an HTML tag(`styled.div`).

```jsx live
const DivWithoutShorthand = styled('div')`
  color: green;
`

const DivWithShorthand = styled.div`
  color: hotpink;
`

render(
  <DivWithoutShorthand>
    This is green. <DivWithShorthand>This is hotpink.</DivWithShorthand>
  </DivWithoutShorthand>
)
```

### Object styles

```jsx live
import styled from 'react-emotion'

const H1 = styled.h1(
  {
    fontSize: 20
  },
  props => ({ color: props.color })
)

render(
  <H1 color="lightgreen">
    This is lightgreen.
  </H1>
)
```

This API was inspired by [glamorous](https://github.com/paypal/glamorous).

### withConfig is not a function error

This error is caused by using the shorthand syntax for styled such as `styled.div` without the Babel plugin. To fix this, [install `babel-plugin-emotion`](/docs/babel.md)
