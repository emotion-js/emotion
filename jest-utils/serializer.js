const css = require('css')
const { sheet } = require('../src/index')

const serializer = { test, print }

module.exports = serializer

function test (val) {
  return (
    val && !val.withStyles && val.$$typeof === Symbol.for('react.test.json')
  )
}

function print (val, printer) {
  const selectors = getSelectors(val)
  const styles = getStyles(selectors)
  val.withStyles = true
  const printedVal = printer(val)
  if (styles) {
    return `${styles}\n\n${printedVal}`
  } else {
    return printedVal
  }
}

function getSelectors (node) {
  let selectors = []
  if (node.children && node.children.reduce) {
    selectors = node.children.reduce(
      (acc, child) => acc.concat(getSelectors(child)),
      []
    )
  }
  if (node.props) {
    return getSelectorsFromProps(selectors, node.props)
  }
  return selectors
}

function getSelectorsFromProps (selectors, props) {
  const className = props.className || props.class
  if (className) {
    selectors = selectors.concat(
      className.toString().split(' ').map(cn => `.${cn}`)
    )
  }
  const dataProps = Object.keys(props).reduce((dProps, key) => {
    if (key.startsWith('data-')) {
      dProps.push(`[${key}]`)
    }
    return dProps
  }, [])
  if (dataProps.length) {
    selectors = selectors.concat(dataProps)
  }
  return selectors
}

function getStyles (nodeSelectors) {
  const styles = sheet.tags
    .map(tag => /* istanbul ignore next */ tag.textContent || '')
    .join('\n')

  const ast = css.parse(styles)
  const rules = ast.stylesheet.rules.filter(filter)
  const mediaQueries = getMediaQueries(ast, filter)

  ast.stylesheet.rules = [...rules, ...mediaQueries]

  const ret = css.stringify(ast)
  return ret

  function filter (rule) {
    if (rule.type === 'rule') {
      const mappedSelectors = rule.selectors.reduce((accum, s) => {
        return accum.concat(s.split('.').filter(Boolean).map(s => `.${s}`))
      }, [])

      return mappedSelectors.some(selector => {
        const baseSelector = selector.split(/:| /)[0]
        return nodeSelectors.includes(baseSelector)
      })
    }
    return false
  }
}

function getMediaQueries (ast, filter) {
  return ast.stylesheet.rules
    .filter(rule => rule.type === 'media')
    .reduce((acc, mediaQuery) => {
      mediaQuery.rules = mediaQuery.rules.filter(filter)

      if (mediaQuery.rules.length) {
        return acc.concat(mediaQuery)
      }

      return acc
    }, [])
}
