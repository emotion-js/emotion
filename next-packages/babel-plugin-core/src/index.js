// @flow
import cssMacro, { transformCssCallExpression } from './css-macro'
import styledMacro from './styled-macro'
import { createPlugin } from './create-plugin'
import { addDefault } from '@babel/helper-module-imports'
import { getSourceMap } from '@emotion/babel-utils'

export const macros = {
  css: cssMacro,
  styled: styledMacro
}

export const appendStringToExpressions = (
  expressions: Array<*>,
  string: string,
  t: *
) => {
  if (!string) {
    return expressions
  }
  if (t.isStringLiteral(expressions[expressions.length - 1])) {
    expressions[expressions.length - 1].value += string
  } else {
    expressions.push(t.stringLiteral(string))
  }
  return expressions
}

let emotionCoreMacroThatsNotARealMacro = ({ references, state, babel }) => {
  Object.keys(references).forEach(refKey => {
    if (refKey === 'css') {
      references[refKey].forEach(path => {
        transformCssCallExpression({ babel, state, path: path.parentPath })
      })
    }
  })
}
emotionCoreMacroThatsNotARealMacro.keepImport = true

export default createPlugin(
  {
    '@emotion/css': cssMacro,
    '@emotion/styled': styledMacro,
    '@emotion/core': emotionCoreMacroThatsNotARealMacro
  },
  babel => {
    let t = babel.types
    return {
      Program(path: *, state: *) {
        if (state.opts.jsx === undefined) {
          for (const node of path.node.body) {
            if (
              t.isImportDeclaration(node) &&
              node.source.value === '@emotion/core'
            ) {
              state.transformCssProp = true
              break
            }
          }
        } else {
          state.transformCssProp = state.opts.jsx
        }
        if (state.opts.sourceMap) {
          state.emotionSourceMap = true
        }
      },
      JSXAttribute(path: *, state: *) {
        if (path.node.name.name !== 'css' && state.transformCssProp) {
          return
        }

        if (
          t.isJSXExpressionContainer(path.node.value) &&
          (t.isObjectExpression(path.node.value.expression) ||
            t.isArrayExpression(path.node.value.expression))
        ) {
          let expressionPath = path.get('value.expression')
          if (expressionPath.isPure()) {
            if (!state.cssIdentifier) {
              state.cssIdentifier = addDefault(path, '@emotion/css', {
                nameHint: 'css'
              })
            }
            expressionPath.replaceWith(
              t.callExpression(
                t.cloneDeep(state.cssIdentifier),
                [
                  path.node.value.expression,
                  state.emotionSourceMap &&
                    path.node.loc !== undefined &&
                    t.stringLiteral(getSourceMap(path.node.loc.start, state))
                ].filter(Boolean)
              )
            )
            transformCssCallExpression({
              babel,
              state,
              path: expressionPath
            })
          }
        }
      }
    }
  }
)
