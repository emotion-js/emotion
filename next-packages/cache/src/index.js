// @flow
import { StyleSheet } from '@emotion/sheet'
import { isBrowser, type CSSContextType } from '@emotion/utils'
import Stylis from '@emotion/stylis'
import ruleSheetPlugin from './rule-sheet'
import type { StylisPlugin } from './types'

export type PrefixOption =
  | boolean
  | ((key: string, value: string, context: 1 | 2 | 3) => boolean)

type StylisPlugins = StylisPlugin[] | StylisPlugin

export type Options = {
  nonce?: string,
  stylisPlugins?: StylisPlugins,
  prefix?: PrefixOption,
  key?: string,
  container?: HTMLElement
}

let createCache = (options?: Options): CSSContextType => {
  if (options === undefined) options = {}
  let key = options.key || 'css'
  let stylisOptions

  if (options.prefix !== undefined) {
    stylisOptions = {
      prefix: options.prefix
    }
  }

  let stylis = new Stylis(stylisOptions)

  stylis.use(options.stylisPlugins)(ruleSheetPlugin)

  if (process.env.NODE_ENV !== 'production') {
    // $FlowFixMe
    if (/[^a-z-]/.test(key)) {
      throw new Error(
        `Emotion key must only contain lower case alphabetical characters and - but "${key}" was passed`
      )
    }
    let sourceMapRegEx = /\/\*#\ssourceMappingURL=data:application\/json;\S+\s+\*\//
    let currentSourceMap
    stylis.use((context, content, selectors) => {
      switch (context) {
        case -1: {
          let result = sourceMapRegEx.exec(content)
          if (result) {
            currentSourceMap = result[0]
          }
          break
        }
        case 2: {
          for (let i = 0, len = selectors.length; len > i; i++) {
            // :last-child isn't included here since it's safe
            // because a style element will never be the last element
            let match = selectors[i].match(/:(first|nth|nth-last)-child/)
            if (match !== null) {
              console.error(
                `The pseudo class "${
                  match[1]
                }" is potentially unsafe when doing server-side rendering. Try changing it to "${
                  match[1]
                }-of-type"`
              )
            }
          }
          break
        }
        case -2: {
          if (currentSourceMap) {
            content.forEach((rule, i) => {
              content[i] = rule + currentSourceMap
            })

            currentSourceMap = ''
          }
        }
      }
    })
  }
  let inserted = {}

  if (isBrowser) {
    // $FlowFixMe
    let parent: HTMLElement = options.container || document.head

    const nodes = document.querySelectorAll(`style[data-emotion-${key}]`)

    Array.prototype.forEach.call(nodes, (node: HTMLStyleElement) => {
      const attrib = node.getAttribute(`data-emotion-${key}`)
      // $FlowFixMe
      attrib.split(' ').forEach(id => {
        inserted[id] = true
      })
      if (node.parentNode !== parent) {
        parent.appendChild(node)
      }
    })
  }

  const context: CSSContextType = {
    stylis,
    key,
    sheet: new StyleSheet({
      key,
      container: options.container,
      nonce: options.nonce
    }),
    inserted,
    registered: {},
    theme: {}
  }
  return context
}

export default createCache
