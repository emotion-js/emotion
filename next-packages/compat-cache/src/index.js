// @flow
import createCache from '@emotion/cache'

let createCompatCache = (emotion: *) => {
  // $FlowFixMe
  let cache = createCache(emotion.sheet.opts)
  cache.inserted = emotion.caches.inserted
  cache.registered = emotion.caches.registered
  cache.compat = true
  return cache
}

export default createCompatCache
