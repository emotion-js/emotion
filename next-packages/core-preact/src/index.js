// an empty file since this is replaced with @emotion/core but with react aliased to our react mock
