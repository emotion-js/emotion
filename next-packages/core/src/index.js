// @flow
export { withCSSContext, Provider } from './context'
export { jsx } from './jsx'
export { Global } from './global'
export { keyframes } from './keyframes'
export { default as css } from '@emotion/css'
