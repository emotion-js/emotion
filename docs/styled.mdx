---
title: 'Styled Components'
---

`styled` is a way to create React components that have styles attached to them. It's available from [@emotion/styled](/packages/@emotion/styled). `styled` was heavily inspired by [styled-components](https://www.styled-components.com/) and [glamorous](https://glamorous.rocks/)

### Styling elements and components

`styled` is very similar to `css` except you call it with an html tag or React component and then call that with a template literal for string styles or a regular function call for object styles.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: turquoise;
`

render(<Button>This my button component.</Button>)
```

### Changing based on props

Any interpolations or arguments that are functions in `styled` are called with `props`, this allows you to change the styles of a component based on the props.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: ${props =>
    props.primary ? 'hotpink' : 'turquoise'};
`

const Container = styled.div(props => ({
  display: 'flex',
  flexDirection: props.column && 'column'
}))

render(
  <Container column>
    <Button>This is a regular button.</Button>
    <Button primary>This is a primary button.</Button>
  </Container>
)
```

### Styling any component

`styled` can style any component as long as it accepts a `className` prop.

```jsx
// @live
import styled from '@emotion/styled'
const Basic = ({ className }) => (
  <div className={className}>Some text</div>
)

const Fancy = styled(Basic)`
  color: hotpink;
`

render(<Fancy />)
```

### Change the rendered tag using `withComponent`

Sometimes you want to create some styles with one component but then use those styles again with another component, the `withComponent` method can be used for this. This was inspired by [styled-components' `withComponent`](https://www.styled-components.com/docs/api#withcomponent).

```jsx
// @live
import styled from '@emotion/styled'

const Section = styled.section`
  background: #333;
`
// this component has the same styles as Section but it renders an aside
const Aside = Section.withComponent('aside')
render(
  <div>
    <Section>This is a section</Section>
    <Aside>This is an an aside</Aside>
  </div>
)
```

### Targeting another emotion component

Similar to [styled-components](https://www.styled-components.com/docs/faqs#can-i-refer-to-other-components), emotion allows for emotion components to be targeted like regular CSS selectors when using [babel-plugin-emotion](/packages/babel-plugin-emotion.md).

```jsx
// @live
import styled from '@emotion/styled'

const Child = styled.div`
  color: red;
`

const Parent = styled.div`
  ${Child} {
    color: green;
  }
`
render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

Component selectors can also be used with object styles.

```jsx
// @live
import styled from '@emotion/styled'

const Child = styled.div({
  color: 'red'
})

const Parent = styled.div({
  [Child]: {
    color: 'green'
  }
})

render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

### Object styles

```jsx
// @live
import styled from '@emotion/styled'

const H1 = styled.h1(
  {
    fontSize: 20
  },
  props => ({ color: props.color })
)

render(<H1 color="lightgreen">This is lightgreen.</H1>)
```

This API was inspired by [glamorous](https://github.com/paypal/glamorous). ❤️

### Customizing prop forwarding

By default, Emotion passes all props to custom components and only props that are valid html attributes for string tags. You can customise this by passing a custom `shouldForwardProp` function. You can also use `@emotion/is-prop-valid` (which is used by emotion internally) to filter out props that are not valid as html attributes.

```jsx
// @live
import isPropValid from '@emotion/is-prop-valid'
import styled from '@emotion/styled'

const H1 = styled('h1', {
  shouldForwardProp: prop =>
    isPropValid(prop) && prop !== 'color'
})(props => ({
  color: 'hotpink'
}))

render(<H1 color="lightgreen">This is lightgreen.</H1>)
```

### Composing dynamic styles

You can create dynamic styles that are based on props and use them in styles.

```jsx
// @live
import styled from '@emotion/styled'
import { css } from '@emotion/core'

const dynamicStyle = props =>
  css`
    color: ${props.color};
  `

const Container = styled.div`
  ${dynamicStyle};
`
render(
  <Container color="lightgreen">
    This is lightgreen.
  </Container>
)
```

### `as` prop

To use styles from a styled component but change the element that's rendered, you can use the `as` prop.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: hotpink;
`

render(
  <Button
    as="a"
    href="https://github.com/emotion-js/emotion"
  >
    Emotion on GitHub
  </Button>
)
```

This API was inspired by [styled-components](https://www.styled-components.com). ❤️

The `as` prop is only used by styled when it's not forwarded to the underlying element. By default, this means that the `as` prop is used for html tags and forwarded for components. To change this, you can pass a custom [`shouldForwardProp`](#customizing-prop-forwarding) which returns `true` for `'as'` to forward it or returns `false` for `'as'` to use it and not forward it.

### Nesting components

We can nest selectors using `&`:

```jsx
// @live

import styled from '@emotion/styled'

const Example = styled('span')`
  color: lightgreen;
  & > a {
    color: hotpink;
  }
`

render(
  <Example>
    This is <a>nested</a>.
  </Example>
)
```
