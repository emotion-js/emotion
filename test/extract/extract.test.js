/* eslint-disable jsx-quotes,no-useless-escape,no-template-curly-in-string */
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { basename } from 'path'
import serializer from 'jest-glamor-react'
import { injectGlobal, css, sheet } from '../../src'
import styled from '../../src/react'

expect.addSnapshotSerializer(serializer(sheet))

describe('styled', () => {
  test('no dynamic', () => {
    const H1 = styled.h1`font-size: 12px;`

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('basic render nested', () => {
    const H1 = styled.h1`
      font-size: 20px;
      & span {
        color: blue;
        
        &:hover {
          color: green;
          
          &:after {
            content: "after";
          }
        }
      }
    `
    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('name', () => {
    const fontSize = '20px'
    const H1 = styled.h1`
      name: FancyH1;
      font-size: 38px;
    `

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('injectGlobal', () => {
    injectGlobal`
      html {
        background: pink;
      }
    `
  })
  test('css', () => {
    expect(css`
      font-family: sans-serif;
      color: yellow;
      background-color: purple;
    `)
  })
  test('writes the correct css', () => {
    const filenameArr = basename(__filename).split('.')
    filenameArr.pop()
    filenameArr.push('emotion', 'css')
    const cssFilename = filenameArr.join('.')
    expect(global.mockedCssImports[cssFilename]).toMatchSnapshot()
  })
})
