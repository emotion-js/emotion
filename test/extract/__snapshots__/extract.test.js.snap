// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled attr 1`] = `
.vars-1sp61ak {
  --css-H1-15vt02v-0: 48;
  --css-H1-15vt02v-1: 4rem;
}

<h1
  className="css-H1-15vt02v vars-1sp61ak"
  fontSize={48}
/>
`;

exports[`styled basic render 1`] = `
.vars-ch9j2 {
  --css-H1-1t8i2zo-0: 20px;
}

<h1
  className="css-H1-1t8i2zo vars-ch9j2"
>
  hello world
</h1>
`;

exports[`styled call expression 1`] = `
.vars-57j4qc {
  --css-H1-rs9k70-0: 20px;
}

<h1
  className="css-H1-rs9k70 vars-57j4qc legacy__class"
>
  hello world
</h1>
`;

exports[`styled composes 1`] = `
.vars-1m8nl4c {
  --css-H1-1pnnqpk-0: 20px;
}

<h1
  className="css-H1-1pnnqpk css-cls2-y5rq6p css-cls1-cy897j vars-1m8nl4c css-H2-idm3bz legacy__class"
  scale={2}
>
  hello world
</h1>
`;

exports[`styled composition 1`] = `
.vars-57j4qc {
  --css-H1-rs9k70-0: 20px;
}

.vars-1u8m2k4 {
  --css-H2-vxb7tq-0: 13.333333333333334px;
}

<h1
  className="css-H1-rs9k70 vars-57j4qc css-H2-vxb7tq vars-1u8m2k4 legacy__class"
>
  hello world
</h1>
`;

exports[`styled function in expression 1`] = `
.vars-57j4qc {
  --css-H1-rs9k70-0: 20px;
}

.vars-6khrnp {
  --css-H2-vxb7tq-0: NaN;
}

<h1
  className="css-H1-rs9k70 vars-57j4qc css-H2-vxb7tq vars-6khrnp legacy__class"
  scale={2}
>
  hello world
</h1>
`;

exports[`styled higher order component 1`] = `
.vars-12p149b {
  --css-Content-rs9k70-0: 20px;
}

<div
  className="css-Content-rs9k70 vars-12p149b css-onyx-13ds1s2 css-squirtle-blue-bg-135ox65"
/>
`;

exports[`styled name 1`] = `
.vars-1vv3got {
  --css-FancyH1-1o9p8a2-0: 20px;
}

<h1
  className="css-FancyH1-1o9p8a2 vars-1vv3got"
>
  hello world
</h1>
`;

exports[`styled no dynamic 1`] = `
<h1
  className="css-H1-ijh7uz"
>
  hello world
</h1>
`;

exports[`styled writes the correct css 1`] = `
".css-H1-ijh7uz { font-size: 12px; }
.css-H1-1t8i2zo { font-size: var(--css-H1-1t8i2zo-0); }
.css-FancyH1-1o9p8a2 {
      font-size: var(--css-FancyH1-1o9p8a2-0); }
.css-H1-nt9mdf { font-size: attr(fontSize);
      margin: attr(margin rem, 4); }
.css-H1-rs9k70 { font-size: var(--css-H1-rs9k70-0); }
.css-H1-rs9k70 { font-size: var(--css-H1-rs9k70-0); }
.css-H2-vxb7tq { font-size: var(--css-H2-vxb7tq-0) }
.css-H1-rs9k70 { font-size: var(--css-H1-rs9k70-0); }
.css-H2-vxb7tq { font-size: var(--css-H2-vxb7tq-0) }
.css-cls1-cy897j { color: blue; }
.css-cls2-y5rq6p {
      height: 64px; }
.css-H1-1pnnqpk {
      font-size: var(--css-H1-1pnnqpk-1); }
.css-H2-idm3bz { font-size: 32px; }
.css-Content-rs9k70 { font-size: var(--css-Content-rs9k70-0); }
.css-squirtle-blue-bg-135ox65 {
      background-color: #7FC8D6; }
.css-onyx-13ds1s2 {
        background-color: '#343a40';
        -webkit-flex-direction: column;
        -ms-flex-direction: column;
        -webkit-box-orient: vertical;
        -webkit-box-direction: normal;
        flex-direction: column; }
html {
        background: pink;
      }
.css-14yvnlz { font-family: sans-serif;
      color: yellow;
      background-color: purple; }"
`;
